Android: Expose textBreakStrategy on Text and TextInput

rigdern opened the pull request:
Android has a text API called breakStrategy for controlling how paragraphs are broken up into lines. For example, some modes support automatically hyphenating words so a word can be split across lines while others do not.

One source of complexity is that Android provides different defaults for `breakStrategy` for `TextView` vs `EditText`. `TextView`'s default is `BREAK_STRATEGY_HIGH_QUALITY` while `EditText`'s default is `BREAK_STRATEGY_SIMPLE`.

In addition to exposing `textBreakStrategy`, this change also fixes a couple of rendering glitches with `Text` and `TextInput`. `TextView` and `EditText` have different default values for `breakStrategy` and `hyphenationFrequency` than `StaticLayout`. Consequently, we were using different parameters for measuring and rendering. Whenever measuring and rendering parameters are inconsistent, it can result in visual glitches such as the text taking up too much space or being clipped.

This change fixes these inconsistencies by setting `breakStrategy` and `hyphenationFrequency` to the appropriate values during measuring.

**Test plan (required)**

Verified the `textBreakStrategy` prop works in a test app. Also, my team's app uses this change.

Adam Comella
Microsoft Corp.

leeight commented on Libraries/Components/TextInput/TextInput.js:
```diff
@@ -316,6 +316,12 @@ const TextInput = React.createClass({
+     * The default value is `simple`
+     * @platform android
+     */
+    textBreakStrategy: PropTypes.string,
```
use enum type?

mkonicek commented:
Looks great! Just a question about the default value:

It looks like Android's default text break strategy is `SIMPLE`:
https://github.com/android/platform_frameworks_base/blob/master/core/java/android/widget/TextView.java#L743

What do you think of defaulting to simple in React Native? This way:
- We're not changing the behavior for everyone
- We're not introducing a potential perf hit for everyone. I tried searching the Android docs for how expensive using the `HIGH_QUALITY` text break strategy is but couldn't find anything.

mkonicek commented:
I fixed two tiny things in the docs, hope you don't mind: https://github.com/facebook/react-native/pull/11007/commits

rigdern commented:
@mkonicek 

> I fixed two tiny things in the docs, hope you don't mind: https://github.com/facebook/react-native/pull/11007/commits

That's fine with me. Thanks!

> It looks like Android's default text break strategy is `SIMPLE`: https://github.com/android/platform_frameworks_base/blob/master/core/java/android/widget/TextView.java#L743. What do you think of defaulting to simple in React Native?

Despite the code you've referenced, I'm pretty sure `TextView` defaults to `BREAK_STRATEGY_HIGH_QUALITY` and `EditText` defaults to `BREAK_STRATEGY_SIMPLE`. I implemented the PR such that RN matches these defaults. Here's my evidence:
  - [The Android docs for `setBreakStrategy`](https://developer.android.com/reference/android/widget/TextView.html#setBreakStrategy(int)) indicate that these values are the defaults.
  - `styles.xml` in the Android source agrees with the docs (my Android experience is limited so I'm not sure if/when this file is used):
    - [TextView's `breakStrategy`](https://github.com/android/platform_frameworks_base/blob/c6c005ef5b927be1b49b4d18a0b887b571fb3ca4/core/res/res/values/styles.xml#L502)
    - [EditText's `breakStrategy`](https://github.com/android/platform_frameworks_base/blob/c6c005ef5b927be1b49b4d18a0b887b571fb3ca4/core/res/res/values/styles.xml#L534)
  - [Here's the commit](https://github.com/android/platform_frameworks_base/commit/39b4db73c3340ff955f67e4e5318159d19d1ab3a) which introduced the `breakStrategy` feature. The defaults described in its commit message match the currently stated defaults from the docs.

mkonicek commented:
Don't know how I missed this in the Android docs, sorry!

> The default value for TextView is BREAK_STRATEGY_HIGH_QUALITY

@facebook-github-bot shipit

mkonicek commented:
There is a merge conflict in TextInput.js now, sorry about that! Can you rebase please?

rigdern commented:
@mkonicek I rebased my branch on master and pushed. One of the conflicts was from one of my other PRs.

mkonicek commented:
Thanks @rigdern!

The [Circle CI test failure](https://circleci.com/gh/facebook/react-native/13558?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) looks related:

`TextInputTestCase` failed:

    E/AndroidRuntime( 2456): Process: com.facebook.react.tests, PID: 2456
    E/AndroidRuntime( 2456): java.lang.NoSuchMethodError: android.widget.EditText.getBreakStrategy
    E/AndroidRuntime( 2456): 	at com.facebook.react.views.textinput.ReactTextInputShadowNode.measure(ReactTextInputShadowNode.java:109)
    ...

I'm hoping the documented way to run tests locally still works? http://facebook.github.io/react-native/docs/testing.html#integration-tests-android

lacker commented:
Travis failure here looks like garbage, I'll rerun it. AFAICT only flaky tests and merge problems are preventing this from being accepted now....

rigdern commented:
@lacker You mentioned merge problems so I just rebased on master and pushed. Is there anything else you need from me before this can be merged?

lacker commented:
No I will merge as soon as Circle/Travis look ok.

rigdern commented:
@lacker Travis passed but CircleCI is still pending for some reason. It has been 2 days since I pushed so it seems like something went wrong and CircleCI will never run on my change. Is there a way to kick a CircleCI build? Should I just push again?

Here's the list of all of the CircleCI builds for this PR: https://circleci.com/gh/facebook/react-native/tree/pull%2F11007. You can see the last one was from a change I pushed 10 days ago.

lacker commented:
Argh. Well I think one of my H1 2017 goals needs to be making this automatic testing setup suck less. Yeah, can you try just pushing again?

rigdern commented:
@lacker yay, CircleCI finally ran on this PR. Both Travis and Circle are now done.

lacker commented:
OK it looks like the FB-internal continuous integration is failing because there is a test which explicitly constructs a `ReactTextUpdate` using the old signature. I'd like to just kind of move on with this though. How about adding a constructor which keeps the old signature as well?

rigdern commented:
@lacker The iOS build failed on Travis but this PR adds an Android-only feature. I suspect the Travis failure is unrelated to my change. How do you want to handle this?

lacker commented:
Yeah Travis was broken yesterday. Our sync script insists that the tests pass so let me just rerun travis and once that works I will land this.
